File: backend/app/schemas/report.py
from pydantic import BaseModel
from typing import Optional, List
from datetime import datetime, date

# ============================================
# Common Areas Usage Report
# ============================================

class CommonAreaUsageStats(BaseModel):
    """Statistics for a single common area"""
    common_area_id: str
    common_area_name: str
    total_reservations: int
    total_hours_reserved: float
    most_popular_day: Optional[str] = None
    average_duration_hours: Optional[float] = None

    class Config:
        from_attributes = True

class CommonAreasUsageReport(BaseModel):
    """Report of common areas usage"""
    start_date: date
    end_date: date
    total_reservations: int
    total_areas_used: int
    areas_stats: List[CommonAreaUsageStats]

# ============================================
# Reservations Report
# ============================================

class ReservationReportItem(BaseModel):
    """Single reservation item in report"""
    id: str
    common_area_name: str
    user_name: str
    user_email: str
    unit_number: Optional[str] = None
    start_time: datetime
    end_time: datetime
    status: str
    duration_hours: float
    created_at: datetime

    class Config:
        from_attributes = True

class ReservationsReport(BaseModel):
    """Report of reservations by period"""
    start_date: date
    end_date: date
    total_reservations: int
    confirmed_count: int
    pending_count: int
    cancelled_count: int
    total_hours_reserved: float
    reservations: List[ReservationReportItem]

# ============================================
# General Statistics
# ============================================

class GeneralStats(BaseModel):
    """General statistics for dashboard"""
    total_users: int
    total_units: int
    total_common_areas: int
    total_reservations_this_month: int
    total_reservations_last_month: int
    most_used_area: Optional[str] = None
    least_used_area: Optional[str] = None

File: frontend/admin/src/types/report.ts
// ============================================
// Common Areas Usage Report
// ============================================

export interface CommonAreaUsageStats {
    common_area_id: string
    common_area_name: string
    total_reservations: number
    total_hours_reserved: number
    most_popular_day: string | null
    average_duration_hours: number | null
}

export interface CommonAreasUsageReport {
    start_date: string // YYYY-MM-DD
    end_date: string // YYYY-MM-DD
    total_reservations: number
    total_areas_used: number
    areas_stats: CommonAreaUsageStats[]
}

// ============================================
// Reservations Report
// ============================================

export interface ReservationReportItem {
    id: string
    common_area_name: string
    user_name: string
    user_email: string
    unit_number: string | null
    start_time: string // ISO 8601
    end_time: string // ISO 8601
    status: string
    duration_hours: number
    created_at: string // ISO 8601
}

export interface ReservationsReport {
    start_date: string // YYYY-MM-DD
    end_date: string // YYYY-MM-DD
    total_reservations: number
    confirmed_count: number
    pending_count: number
    cancelled_count: number
    total_hours_reserved: number
    reservations: ReservationReportItem[]
}

// ============================================
// Report Filters
// ============================================

export interface ReportFilters {
    start_date?: string
    end_date?: string
    status?: 'confirmed' | 'pending' | 'cancelled'
    common_area_id?: string
}
